### PR TITLE
Feature/svcs 452 dry run

### PR DESCRIPTION
**Pull request by arnedubois**

### CODE REVIEW

File: src/commands/data/schemas/util/syncSchema.ts [hunk 1]
```diff
@@ -359,3 +297,185 @@ function deepDiff(object: any, other: any) {
...
+  ));
+
+  return {
+    toAdd,
+    toRemove,
+    toUpdate,
+  };
+}
+
+function compareArraysByName(localSchema: any[], cloudSchema: any[]) {
```
Comment (chad-pelly) [inline src/commands/data/schemas/util/syncSchema.ts:404]: Personally I would make this a little more 'readable'

```
function compareArraysByName(localSchema: any[], cloudSchema: any[]) {
  const toAdd: any[] = _.differenceBy(localSchema, cloudSchema, 'name');
  const toRemove: any[] = _.differenceBy(cloudSchema, localSchema, 'name');

  // Add quite a long comment here about what this logical grouping does
  const missingFields = _.differenceBy(localSchema, toAdd, 'name');
  const excessFields = _.differenceBy(cloudSchema, toRemove, 'name');
  const toUpdate: any = _.differenceWith(
    missingFields,
    excessFields,
    (targetTransition, currentTransition) => _.isEqual(
      _.omit(targetTransition, ['id']),
      _.omit(currentTransition, ['id'])
    )
  );

  return {
    toAdd,
    toRemove,
    toUpdate,
  };
}
```

File: src/commands/data/schemas/util/syncSchema.ts [hunk 2]
```diff
@@ -359,3 +297,185 @@ function deepDiff(object: any, other: any) {
...
+  );
+
+  // calculate excess properties
+  const toRemove = _.difference(
+    Object.keys(cloudSchema),
+    Object.keys(localSchema)
+  );
+
+  // calculate properties which have changed
+  const toUpdate: any = Object.keys(deepDiff(
```
Comment (chad-pelly) [inline src/commands/data/schemas/util/syncSchema.ts:392]: This is of type `string[]` right?

File: src/commands/data/schemas/util/syncSchema.ts [hunk 3]
```diff
@@ -359,3 +297,185 @@ function deepDiff(object: any, other: any) {
...
+      cloudSchema[key]
+    );
+  }
+  return compareSchemas(
+    localSchema[key],
+    cloudSchema[key]
+  );
+}
+
+function compareSchemas(localSchema: any, cloudSchema: any) {
```
Comment (chad-pelly) [inline src/commands/data/schemas/util/syncSchema.ts:379]: For readability I would do something like this

```
function compareSchemas(localSchema: any, cloudSchema: any) {
  // Calculate the properties of the schemas that have been added or removed
  const toAdd = _.difference(Object.keys(localSchema), Object.keys(cloudSchema));
  const toRemove = _.difference(Object.keys(cloudSchema), Object.keys(localSchema));

  // Calculate the properties of the schemas that have been updated
  const existingLocalProperties = _(localSchema).omit(toAdd).value();
  const existingCloudProperties = _(cloudSchema).omit(toRemove).value();
  const schemaDiff = deepDiff(existingLocalProperties, existingCloudProperties);

  const toUpdate = Object.keys(schemaDiff);

  return {
    toAdd,
    toRemove,
    toUpdate,
  };
}
```

File: src/commands/data/schemas/util/syncSchema.ts [hunk 4]
```diff
@@ -359,3 +297,185 @@ function deepDiff(object: any, other: any) {
...
+}) {
+  const changes = compareArraysByName(newIndexes, removedIndexes);
+  logChanges(`Schema ${schema.name} - Indexes`, {
+    toAdd: changes.toAdd.map(v => v.name),
+    toRemove: changes.toRemove.map(v => v.name),
+    toUpdate: changes.toUpdate.map(v => v.name),
+  });
+}
+
+export function compareSchemaKey(localSchema: any, cloudSchema: any, key: string) {
```
Comment (chad-pelly) [inline src/commands/data/schemas/util/syncSchema.ts:355]: Just for readability without changing any logic at all

```
export function compareSchemaKey(localSchema: any, cloudSchema: any, key: string) {
  const isInvalidLocalProperty = !localSchema[key];
  const isInvalidCloudProperty = !cloudSchema[key];

  if (isInvalidLocalProperty && isInvalidCloudProperty) {
    return {
      toAdd: [],
      toRemove: [],
      toUpdate: [],
    };
  }

  const isLocalPropertyArray = Array.isArray(localSchema[key]);
  const isCloudPropertyArray = Array.isArray(cloudSchema[key]) && isInvalidLocalProperty;

  if (isLocalPropertyArray || isCloudPropertyArray) {
    return compareArraysByName(localSchema[key], cloudSchema[key]);
  }

  return compareSchemas(localSchema[key], cloudSchema[key]);
}
```

File: src/commands/data/schemas/util/syncSchema.ts [hunk 5]
```diff
@@ -359,3 +297,185 @@ function deepDiff(object: any, other: any) {
...
+  const changedKeys = Object.keys(updatedValues);
+  console.group(`Schema ${cloudSchema.name} - Root attributes`);
+  if (changedKeys.length < 1) { console.log('No update required.'); }
+  for (const key of changedKeys) {
+    console.log(`${chalk.yellow(key)}:\t ${chalk.red(cloudSchema[key])} => ${chalk.green(updatedValues[key])}`);
+  }
+  console.groupEnd();
+}
+
+function logChanges(group: string, { toAdd, toRemove, toUpdate }: {
```
Comment (chad-pelly) [inline src/commands/data/schemas/util/syncSchema.ts:318]: Again just for readability / simplicity

```
type changes = { toAdd: string[]; toRemove: string[]; toUpdate: string[]; };
function reportSchemaChanges(group: string, changes: changes) {
  const { toAdd, toRemove, toUpdate } = changes;

  console.group(group);

  // Should resolve correctly without the > 1 , 0 is falsy right?
  if (toAdd.length && toRemove.length && toUpdate.length) {
    console.log('No update required');
    return;
  }

  // Report changes to the schema
  forEach(toAdd, key => console.log(`Will be added: ${chalk.green(getIdentifier(key))}`));
  forEach(toRemove, key => console.log(`Will be removed: ${chalk.red(getIdentifier(key))}`));
  forEach(toUpdate, key => console.log(`Will be updated: ${chalk.yellow(getIdentifier(key))}`));

  console.groupEnd();
}
```

Comment (chad-pelly) [reply src/commands/data/schemas/util/syncSchema.ts:318]: I would also take this example and apply the change ideology to 'logDifferentRootAttributes' , but it's all opinion based!

File: src/commands/data/schemas/util/syncSchema.ts [hunk 6]
```diff
@@ -359,3 +297,185 @@ function deepDiff(object: any, other: any) {
...
+}
+
+function findTransitionByName(name: string, transitions: any[]) {
+  return _.find(
+    transitions,
+    transition => transition.name === name
+  );
+}
+
+function compareIndexes(localSchema: any, cloudSchema: any) {
```
Comment (chad-pelly) [inline src/commands/data/schemas/util/syncSchema.ts:445]: Twould also be nice if you liked some ideas in my readability examples if you could apply that here, but again it's opinion based.

**Review by chad-pelly — COMMENTED** [review]
GG! Nothing major, a lot of opinions.

File: src/commands/data/schemas/util/syncSchema.ts (hunk 2 above)
Comment (arnedubois) [reply src/commands/data/schemas/util/syncSchema.ts:392]: Yes, likely leftover from a copy paste

**Review by chad-pelly — APPROVED** [review]
LGTM! 🔥

**Pull request merged**